module/VuFind/src/VuFind/Auth/AbstractShibboleth.php

```diff
@@ -0,0 +1,362 @@
+        // attributes since we need to use saveCredentials method to store it):
+        $catPassword = null;
+
+        // Has the user configured attributes to use for populating the user table?
+        foreach ($this->attribsToCheck as $attribute) {
+            if (isset($shib[$attribute])) {
+                $value = $this->getAttribute($request, $shib[$attribute]);
+                if ($attribute == 'email') {
+                    $user->updateEmail($value);
+                } elseif ($attribute == 'cat_username' && isset($shib['prefix'])) {
+                    $user->cat_username = $shib['prefix'] . '.'
+                        . (($value === null) ? '' : $value);
+                } elseif ($attribute == 'cat_password') {
+                    $catPassword = $value;
+                } else {
+                    $user->$attribute = ($value === null) ? '' : $value;
+                }
+            }
+        }
+
+        // Save credentials if applicable. Note that we want to allow empty
+        // passwords (see https://github.com/vufind-org/vufind/pull/532), but
+        // we also want to be careful not to replace a non-blank password with a
+        // blank one in case the auth mechanism fails to provide a password on
+        // an occasion after the user has manually stored one. (For discussion,
+        // see https://github.com/vufind-org/vufind/pull/612). Note that in the
+        // (unlikely) scenario that a password can actually change from non-blank
+        // to blank, additional work may need to be done here.
+        if (!empty($user->cat_username)) {
+            $user->saveCredentials(
+                $user->cat_username,
+                empty($catPassword) ? $user->getCatPassword() : $catPassword
+            );
+        }
+
+        $this->storeShibbolethSession($request);
+
+        // Save and return the user object:
+        $user->save();
+        return $user;
+    }
+
+    /**
+     * Return shibboleth configuration.
+     *
+     * @param Request $request Request object
+     *
+     * @return array
+     */
+    abstract protected function getShibbolethConfiguration($request);
+
+    /**
+     * Extract required user attributes from the configuration.
+     *
+     * @param array $config shibboleth configuration
+     *
+     * @return array      Only username and attribute-related values
+     * @throws AuthException
+     */
+    protected function getRequiredAttributes($config)
+    {
+        // Special case -- store username as-is to establish return array:
+        $sortedUserAttributes = [];
+
+        // Now extract user attribute values:
+        $shib = $this->getConfig()->Shibboleth;
+        foreach ($shib as $key => $value) {
+            if (preg_match("/userattribute_[0-9]{1,}/", $key)) {
+                $valueKey = 'userattribute_value_' . substr($key, 14);
+                $sortedUserAttributes[$value] = $shib['$valueKey'] ?? null;
```

Comment:
This doesn't look right -- putting single quotes around `$valueKey` means that the literal string will be used rather than the value of the variable. Is there even a reason that this had to be changed from object notation to array notation in this situation?

Reply:
...or is the idea that you were supposed to be using the `$config` parameter to this function (which currently appears to be unused) instead of calling `$this->getConfig()`?